app/services/ValidatorPool.scala

```diff
@@ -1,67 +1,136 @@
+    * If no ids are assigned, use all the currently available validators.
+    */
+  def check(id: String, text: String, maybeCategoryIds: Option[List[String]] = None): Future[List[RuleMatch]] = {
+    val categoryIds = maybeCategoryIds match {
+      case None => getCurrentCategories.map { case (_, category) => category.id }
+      case Some(ids) => ids
+    }
+
+    Logger.info(s"Validation job with id: ${id} received. Checking categories: ${categoryIds.mkString(", ")}.")
+
+    val eventualChecks = categoryIds.map { categoryId =>
+      checkForCategory(id, text, categoryId)
+    }
 
-  def checkAllCategories(text: String): Future[ValidatorResponse] = {
-    Logger.info(s"Checking categories ${validators.keys.mkString(", ")}")
-    val eventualChecks = validators.keys.foldLeft[List[Future[ValidatorResponse]]](List.empty)((acc, category) => {
-      acc :+ check(ValidatorRequest(category, text))
-    })
     Future.sequence(eventualChecks).map {
       _.flatten
+    }.map { matches =>
+      Logger.info(s"Validation job with id: ${id} complete.")
+      matches
     }
   }
 
-  def check(request: ValidatorRequest): Future[ValidatorResponse] = {
-    validators.get(request.category) match {
-      case Some(validator) =>
-        Logger.info(s"Validator for category ${request.category} is processing")
-        validator.future.map { validator =>
-          blocking {
-            val result = validator.check(request)
-            Logger.info(s"Validator for category ${request.category} is done")
-            result
+  /**
+    * Add a validator to the pool of validators for the given category.
+    * Replaces a validator that's already present for that category, returning
+    * the replaced validator.
+    */
+  def addValidator(category: Category, validator: Validator): Option[(Category, Validator)] = {
+    Logger.info(s"New instance of validator available of id: ${validator.getId} for category: ${category.id}")
+    validators.put(category.id, (category, validator))
+  }
+
+  def getCurrentCategories: List[(String, Category)] = {
+    val validatorsAndCategories = validators.values.map {
+      case (category, validator) => (validator.getId, category)
+    }.toList
+    validatorsAndCategories
+  }
+
+  def getCurrentRules: List[Rule] = {
+    validators.values.flatMap {
+      case (_, validator) =>  validator.getRules
+    }.toList
+  }
+
+  private def checkForCategory(id: String, text: String, categoryId: String): Future[ValidatorResponse] = {
+    val promise = Promise[ValidatorResponse]()
+    val future = promise.future
+    // Wrap our request in a lambda and add it to the queue, ready to be called.
+    val validate: () => Future[ValidatorResponse] = () => validators.get(categoryId) match {
+      case Some((_, validator)) =>
+        val eventualResult = validator.check(ValidatorRequest(categoryId, text))
+        eventualResult.andThen {
+          case Success(result) => {
```

Comment:
@akash1810 this is where we maintain a reference to our caller as we're sat in the queue -- and where the code gets a little windy!